File: bids-expression-evaluator/environment/app/src/parser.ts

import { Token, TokenType, ASTNode } from './types';
import { tokenize } from './lexer';

class Parser {
  private tokens: Token[];
  private pos: number = 0;

  constructor(tokens: Token[]) {
    this.tokens = tokens;
  }

  private peek(): Token {
    return this.tokens[this.pos];
  }

  private advance(): Token {
    const token = this.tokens[this.pos];
    this.pos++;
    return token;
  }

  private expect(type: TokenType): Token {
    const token = this.peek();
    if (token.type !== type) {
      throw new Error(
        `Expected ${type} but got ${token.type} (${token.value}) at position ${token.position}`
      );
    }
    return this.advance();
  }

  parse(): ASTNode {
    const node = this.parseExpression();
    if (this.peek().type !== TokenType.EOF) {
      throw new Error(`Unexpected token: ${this.peek().type} (${this.peek().value})`);
    }
    return node;
  }

  private parseExpression(): ASTNode {
    return this.parseOr();
  }

  private parseOr(): ASTNode {
    let left = this.parseAnd();
    while (this.peek().type === TokenType.Or) {
      this.advance();
      const right = this.parseAnd();
      left = { type: 'BinaryOp', operator: '||', left, right };
    }
    return left;
  }

  private parseAnd(): ASTNode {
    let left = this.parseEquality();
    while (this.peek().type === TokenType.And) {
      this.advance();
      const right = this.parseEquality();
      left = { type: 'BinaryOp', operator: '&&', left, right };
    }
    return left;
  }

  private parseEquality(): ASTNode {
    let left = this.parseComparison();
    while (
      this.peek().type === TokenType.EqualEqual ||
      this.peek().type === TokenType.BangEqual
    ) {
      const op = this.advance().value;
      const right = this.parseComparison();
      left = { type: 'BinaryOp', operator: op, left, right };
    }
    return left;
  }

  private parseComparison(): ASTNode {
    let left = this.parseAdditive();
    while (
      this.peek().type === TokenType.Less ||
      this.peek().type === TokenType.LessEqual ||
      this.peek().type === TokenType.Greater ||
      this.peek().type === TokenType.GreaterEqual
    ) {
      const op = this.advance().value;
      const right = this.parseAdditive();
      left = { type: 'BinaryOp', operator: op, left, right };
    }
    return left;
  }

  private parseAdditive(): ASTNode {
    let left = this.parseMultiplicative();
    while (
      this.peek().type === TokenType.Plus ||
      this.peek().type === TokenType.Minus
    ) {
      const op = this.advance().value;
      const right = this.parseMultiplicative();
      left = { type: 'BinaryOp', operator: op, left, right };
    }
    return left;
  }

  private parseMultiplicative(): ASTNode {
    let left = this.parseUnary();
    while (
      this.peek().type === TokenType.Star ||
      this.peek().type === TokenType.Slash
    ) {
      const op = this.advance().value;
      const right = this.parseUnary();
      left = { type: 'BinaryOp', operator: op, left, right };
    }
    return left;
  }

  private parseUnary(): ASTNode {
    if (this.peek().type === TokenType.Bang) {
      this.advance();
      const operand = this.parseUnary();
      return { type: 'UnaryOp', operator: '!', operand };
    }
    if (this.peek().type === TokenType.Minus) {
      this.advance();
      const operand = this.parseUnary();
      return { type: 'UnaryOp', operator: '-', operand };
    }
    return this.parsePostfix();
  }

  private parsePostfix(): ASTNode {
    let node = this.parsePrimary();

    while (true) {
      if (this.peek().type === TokenType.Dot) {
        this.advance();
        const prop = this.expect(TokenType.Identifier).value;
        node = { type: 'MemberAccess', object: node, property: prop };
      } else if (this.peek().type === TokenType.LeftBracket) {
        this.advance();
        const index = this.parseExpression();
        this.expect(TokenType.RightBracket);
        node = { type: 'IndexAccess', object: node, index };
      } else {
        break;
      }
    }

    return node;
  }

  private parsePrimary(): ASTNode {
    const token = this.peek();

    switch (token.type) {
      case TokenType.Number: {
        this.advance();
        return { type: 'NumberLiteral', value: parseFloat(token.value) };
      }

      case TokenType.String: {
        this.advance();
        return { type: 'StringLiteral', value: token.value };
      }

      case TokenType.Boolean: {
        this.advance();
        return { type: 'BooleanLiteral', value: token.value === 'true' };
      }

      case TokenType.Null: {
        this.advance();
        return { type: 'NullLiteral' };
      }

      case TokenType.Identifier: {
        const name = this.advance().value;
        // Function call
        if (this.peek().type === TokenType.LeftParen) {
          this.advance(); // consume '('
          const args: ASTNode[] = [];
          if (this.peek().type !== TokenType.RightParen) {
            args.push(this.parseExpression());
            while (this.peek().type === TokenType.Comma) {
              this.advance();
              args.push(this.parseExpression());
            }
          }
          this.expect(TokenType.RightParen);
          return { type: 'FunctionCall', name, args };
        }
        return { type: 'Identifier', name };
      }

      case TokenType.LeftParen: {
        this.advance();
        const expr = this.parseExpression();
        this.expect(TokenType.RightParen);
        return expr;
      }

      case TokenType.LeftBracket: {
        this.advance();
        const elements: ASTNode[] = [];
        if (this.peek().type !== TokenType.RightBracket) {
          elements.push(this.parseExpression());
          while (this.peek().type === TokenType.Comma) {
            this.advance();
            elements.push(this.parseExpression());
          }
        }
        this.expect(TokenType.RightBracket);
        return { type: 'ArrayLiteral', elements };
      }

      default:
        throw new Error(
          `Unexpected token: ${token.type} (${token.value}) at position ${token.position}`
        );
    }
  }
}

export function parse(input: string): ASTNode {
  const tokens = tokenize(input);
  const parser = new Parser(tokens);
  return parser.parse();
}

File: task-db3fcf/environment/app/parser.py
"""Recursive-descent parser for MiniCalc with operator precedence."""

from lexer import TokenType
from ast_nodes import (
    Program, FuncDef, AssignStmt, IfStmt, WhileStmt, ReturnStmt,
    PrintStmt, IntLit, VarRef, BinOp, UnaryOp, FuncCall,
)


class ParseError(Exception):
    pass


class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = 0

    def peek(self):
        return self.tokens[self.pos]

    def advance(self):
        tok = self.tokens[self.pos]
        self.pos += 1
        return tok

    def expect(self, type_):
        tok = self.advance()
        if tok.type != type_:
            raise ParseError(
                f"Expected {type_}, got {tok.type} ({tok.value!r}) at line {tok.line}")
        return tok

    def parse(self):
        functions = []
        main_body = []
        while self.peek().type != TokenType.EOF:
            if self.peek().type == TokenType.FUNC:
                functions.append(self._parse_func())
            else:
                main_body.append(self._parse_stmt())
        return Program(functions, main_body)

    def _parse_func(self):
        self.expect(TokenType.FUNC)
        name = self.expect(TokenType.IDENT).value
        self.expect(TokenType.LPAREN)
        params = []
        if self.peek().type != TokenType.RPAREN:
            params.append(self.expect(TokenType.IDENT).value)
            while self.peek().type == TokenType.COMMA:
                self.advance()
                params.append(self.expect(TokenType.IDENT).value)
        self.expect(TokenType.RPAREN)
        body = self._parse_block()
        return FuncDef(name, params, body)

    def _parse_block(self):
        self.expect(TokenType.LBRACE)
        stmts = []
        while self.peek().type != TokenType.RBRACE:
            stmts.append(self._parse_stmt())
        self.expect(TokenType.RBRACE)
        return stmts

    def _parse_stmt(self):
        tok = self.peek()
        if tok.type == TokenType.IF:
            return self._parse_if()
        if tok.type == TokenType.WHILE:
            return self._parse_while()
        if tok.type == TokenType.RETURN:
            return self._parse_return()
        if tok.type == TokenType.PRINT:
            return self._parse_print()
        if tok.type == TokenType.IDENT:
            return self._parse_assign()
        raise ParseError(f"Unexpected token {tok.type} at line {tok.line}")

    def _parse_if(self):
        self.expect(TokenType.IF)
        self.expect(TokenType.LPAREN)
        cond = self._parse_expr()
        self.expect(TokenType.RPAREN)
        then_body = self._parse_block()
        else_body = None
        if self.peek().type == TokenType.ELSE:
            self.advance()
            if self.peek().type == TokenType.IF:
                else_body = [self._parse_if()]
            else:
                else_body = self._parse_block()
        return IfStmt(cond, then_body, else_body)

    def _parse_while(self):
        self.expect(TokenType.WHILE)
        self.expect(TokenType.LPAREN)
        cond = self._parse_expr()
        self.expect(TokenType.RPAREN)
        body = self._parse_block()
        return WhileStmt(cond, body)

    def _parse_return(self):
        self.expect(TokenType.RETURN)
        expr = self._parse_expr()
        self.expect(TokenType.SEMI)
        return ReturnStmt(expr)

    def _parse_print(self):
        self.expect(TokenType.PRINT)
        exprs = [self._parse_expr()]
        while self.peek().type == TokenType.COMMA:
            self.advance()
            exprs.append(self._parse_expr())
        self.expect(TokenType.SEMI)
        return PrintStmt(exprs)

    def _parse_assign(self):
        name = self.expect(TokenType.IDENT).value
        self.expect(TokenType.ASSIGN)
        expr = self._parse_expr()
        self.expect(TokenType.SEMI)
        return AssignStmt(name, expr)

    # Expression parsing with operator precedence (lowest to highest)

    def _parse_expr(self):
        return self._parse_or()

    def _parse_or(self):
        left = self._parse_and()
        while self.peek().type == TokenType.OR:
            self.advance()
            right = self._parse_and()
            left = BinOp('||', left, right)
        return left

    def _parse_and(self):
        left = self._parse_equality()
        while self.peek().type == TokenType.AND:
            self.advance()
            right = self._parse_equality()
            left = BinOp('&&', left, right)
        return left

    def _parse_equality(self):
        left = self._parse_comparison()
        while self.peek().type in (TokenType.EQ_EQ, TokenType.BANG_EQ):
            op = self.advance().value
            right = self._parse_comparison()
            left = BinOp(op, left, right)
        return left

    def _parse_comparison(self):
        left = self._parse_addition()
        while self.peek().type in (TokenType.LT, TokenType.LE, TokenType.GT, TokenType.GE):
            op = self.advance().value
            right = self._parse_addition()
            left = BinOp(op, left, right)
        return left

    def _parse_addition(self):
        left = self._parse_multiplication()
        while self.peek().type in (TokenType.PLUS, TokenType.MINUS):
            op = self.advance().value
            right = self._parse_multiplication()
            left = BinOp(op, left, right)
        return left

    def _parse_multiplication(self):
        left = self._parse_unary()
        while self.peek().type in (TokenType.STAR, TokenType.SLASH, TokenType.PERCENT):
            op = self.advance().value
            right = self._parse_unary()
            left = BinOp(op, left, right)
        return left

    def _parse_unary(self):
        if self.peek().type == TokenType.MINUS:
            self.advance()
            operand = self._parse_unary()
            return UnaryOp('-', operand)
        if self.peek().type == TokenType.BANG:
            self.advance()
            operand = self._parse_unary()
            return UnaryOp('!', operand)
        return self._parse_primary()

    def _parse_primary(self):
        tok = self.peek()
        if tok.type == TokenType.INT:
            self.advance()
            return IntLit(tok.value)
        if tok.type == TokenType.IDENT:
            self.advance()
            if self.peek().type == TokenType.LPAREN:
                self.advance()
                args = []
                if self.peek().type != TokenType.RPAREN:
                    args.append(self._parse_expr())
                    while self.peek().type == TokenType.COMMA:
                        self.advance()
                        args.append(self._parse_expr())
                self.expect(TokenType.RPAREN)
                return FuncCall(tok.value, args)
            return VarRef(tok.value)
        if tok.type == TokenType.LPAREN:
            self.advance()
            expr = self._parse_expr()
            self.expect(TokenType.RPAREN)
            return expr
        raise ParseError(f"Unexpected token {tok.type} ({tok.value!r}) at line {tok.line}")
